KAFKA-19434: Startup state stores initialization

eduwercamacaro opened the pull request:
Instead of creating Standby tasks from the state directory, we open the
local stores that exists in the state directory.

This resolves the issue raised in [#KAFKA-19434
](https://issues.apache.org/jira/browse/KAFKA-19434), where
store-specific metrics were being duplicated due to tasks being created
in the main thread and then assigned to a StreamThread.

Additionally, since we can now read the offsets from the store during
instance initialization, this clears the way for the implementation of
KIP-1035. As of now, the stores are loading the offsets from the
checkpoint file, but in a later PR, we will read these offsets from the
state store itself.

This PR modifies the behavior of Kafka Streams when initializing. Now
for each pre-existing store on the state directory: We open the store,
read offsets from the checkpoint file and then close it again. The
reason why we open the store is because the store will be responsible
for tracking the offsets and we will deprecate the checkpoint file.

Reviewers: Nikiita Shuplestov<nshupletsov@confluent.io>, Bill
 Bejeck<bbejeck@apache.org>

eduwercamacaro committed:
refactor startup state stores initialization

Nikita-Shupletsov commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -157,9 +159,14 @@ public RocksDBStore(final String name,
         metricsRecorder.init(metricsImpl(stateStoreContext), stateStoreContext.taskId());
-        openDB(stateStoreContext.appConfigs(), stateStoreContext.stateDir());
+        if (!open) {
+            preInit(stateStoreContext);
```
maybe it's not a valid concern, but it looks like this interface would be hard to explain and thus use correctly: there is a two stage init process. sometimes we call preInit(phase one) explicitly, but there are a lot of cases when the author of the store needs to call it themselves in the init(phase two).

as we rely on stateStore.isOpen be true, maybe we should communicate that and follow the same path every time we want to initialize a store?
i.e.
* phase one(preInit/open)
* check if opened
* phase two(init)

it will slightly complicate the implementation on our side, but will make it significantly easier to follow for users

eduwercamacaro replied:
Thank you for your time reviewing this PR.

That is a valid concern. 

As things are right now in this PR, it's not clear when each method in the state store lifecycle is called. This makes things a lot harder for the people who write state stores, including the existing `RocksDBStore`. 

I agree that we need to make the implementation a little harder on our end and a little easier on the specific state stores’ end. 

I was thinking that we could document the `preInit` method and specify that it will be called immediately after the state store is built (in the topology builder), allowing state stores to implement logic for opening resources that will be required during the `init` phase. In the case of the `RocksDBStore`, we will open the store during this phase.

Although I am not familiar with custom state stores, I would assume that they could also open a database connection during this phase. 

Since the state store was already opened during the `preInit` phase, the `init` method can only concentrate on preparing it for use in StreamThread processing (i.e., registering metrics in the process context). This is because the init method will run during the rebalance process.

State store authors would find it easy to implement or evolve their custom implementations with a more clear lifecycle.

Nikita-Shupletsov commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -221,44 +228,23 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
+                    // TODO: we need to pass a proper logPrefix
+                    StateManagerUtil.registerStateStores(log, "", subTopology, stateManager, this, initContext);
+                    for (final StateStore stateStore : subTopology.stateStores()) {
+                        if (!stateStore.isOpen()) {
```
I assume, the rest of stores will be updated later? right now there is only one Store updated: RockDBStore. I opened a random store(InMemoryKeyValueStore) and it sets the open field to true in the init method, so it will start failing here, I think

eduwercamacaro replied:
Yes, this needs to be updated for the other stores.

eduwercamacaro replied:
Thanks @Nikita-Shupletsov 
There are many state stores that needed to be updated with the preInit implementation. I pushed [this commit](https://github.com/apache/kafka/pull/20749/commits/b4cb0ff7bba4218251cdf4490edca0b2afd665d0) to address this comment. 

Let me know what you about it.

Nikita-Shupletsov replied:
@eduwercamacaro 
thanks!
just to confirm: the idea is that we want the owner of a store to be responsible for its proper initialization, right? so if KS hasn't called preInit, they will need to do that themselves during init

Nikita-Shupletsov reviewed (commented):
Thanks a lot for putting it together!
A much better idea than what I had in my PR

eduwercamacaro committed:
merge trunk

eduwercamacaro committed:
Remove rocksdb temp files

eduwercamacaro committed:
Merge branch 'trunk' into refactor-startup-store-initialization

eduwercamacaro committed:
Fix merge conflicts

eduwercamacaro committed:
Replace RocksDBStore#openDB to StateStore#preInit

eduwercamacaro committed:
Remove unused classes

eduwercamacaro committed:
Discover startup stores during topology initialization

eduwercamacaro committed:
Allow closing state stores during preInit phase

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -205,7 +212,8 @@ public ProcessorStateManager(final TaskId taskId,
                                  final Collection<TopicPartition> sourcePartitions,
-                                 final boolean stateUpdaterEnabled) throws ProcessorStateException {
+                                 final boolean stateUpdaterEnabled,
+                                 final boolean startupState) throws ProcessorStateException {
```
What does `startupState` represent, maybe rename to something like `hasStateOnStartup`?

eduwercamacaro replied:
Mostly, this boolean flag controls when to call the `preInit` and the `init` methods within this class. This is because the `ProcessorStateManager` is the one that controls the state store lifecycle. 

But I think it is probably better to create a new method that specifically registers startup states. I made some changes that removes this field by adding a new method in the ProcessorStateManager. see [this commit](https://github.com/apache/kafka/pull/20749/commits/fd93623864f3ae78edab8ac876ded56a495b21ac)

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -220,6 +228,22 @@ public ProcessorStateManager(final TaskId taskId,
         this.checkpointFile = new OffsetCheckpoint(stateDirectory.checkpointFileFor(taskId));
 
         log.debug("Created state store manager for task {}", taskId);
+        this.startupState = new AtomicBoolean(startupState);
```
same would apply to the `AtomicBoolean` variable

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -255,6 +279,10 @@ void assignToStreamThread(final LogContext logContext,
         this.sourcePartitions.addAll(sourcePartitions);
     }
 
+    void reuseState() {
```
What are the semantics around `reuseState`? Seems a little odd that calling `reuse` sets false, probably could use a small comment or a method rename.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -333,25 +333,19 @@ private void closeDirtyAndRevive(final Collection<Task> taskWithChangelogs, fina
 
-    private Map<Task, Set<TopicPartition>> assignStartupTasks(final Map<TaskId, Set<TopicPartition>> tasksToAssign,
-                                                              final String threadLogPrefix,
-                                                              final TopologyMetadata topologyMetadata,
```
Why did these parameters go away?

eduwercamacaro replied:
This PR only removes the `threadLodPrefix` parameter. It now uses the `logPrefix` instance variable. [see this commit](https://github.com/apache/kafka/pull/20749/commits/49e786f4f7379c57366b1660454f148955fed941)

bbejeck reviewed (commented):
Thanks @eduwercamacaro! I made an initial pass

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -230,44 +238,23 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
-                        task.suspend();
-                        task.closeDirty();
+                    final StartupContext initContext = new StartupContext(id, config, stateManager);
+                    // TODO: we need to pass a proper logPrefix
```
instead of `""` we could use `logPrefix()`

eduwercamacaro replied:
I'm not totally sure if that's what we want because the `logPrefix()` method will use a `stream-thread-` as prefix. I think we will need to provide a log prefix when assigning these startup state store to a Task.

eduwercamacaro replied:
I think we need to use a different format for the logPrefix since this method is not being invoked by a StreamThread. Just pushed [this commit ](https://github.com/apache/kafka/pull/20749/commits/f8febe71461adae7288943382a27ba9f91436a46)let me know what you think about it.

bbejeck reviewed (commented):
@eduwercamacaro I've made another pass and confirmed the fix by running a streams application and stopping/starting several times with existing state.  Overall this lgtm
Can you rebase this PR?

Also I think we could use a new test in `KafkaStreamsTelemetryIntegrationTest` that  start/stops/restarts a single instance and confirms the state metrics getting emitted.

eduwercamacaro committed:
Add a new method that register startup stores in a ProcessorStateManager

eduwercamacaro committed:
Merge branch 'trunk' into refactor-startup-store-initialization

eduwercamacaro commented:
@bbejeck Thanks for your time reviewing this. I already rebased this branch.

I believe [this integration test](https://github.com/apache/kafka/pull/20749/files#diff-a8f919d43ec1cebdb136acf94adfbeb20fef0cda0ad5d29b7010e0186b355d71R287) is already covering that case. what do you think?

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/StateStore.java:
```diff
@@ -71,6 +71,8 @@ public interface StateStore {
     void init(final StateStoreContext stateStoreContext, final StateStore root);
 
+    default void preInit(final StateStoreContext stateStoreContext) {}
+
```
Missed this before - what's the functionality of `preInit` vs. `init`

eduwercamacaro replied:
This is the part of KIP-1035 that needs to be changed so that this new method can be added to the StateStore interface. 

`preInit` essentially adds a new callback for StateStores that allows us to initialize resources before the StateStore is assigned to a `StreamThread` via `StateStore#init`. For the case of RocksDBStateStore, we will load current offsets from the ColumnFamily.

On this PR, we need this new method because it allows the StateDirectory class to open the state stores without assigning them to a StreamThread.

bbejeck reviewed (commented):
Thanks again for the PR @eduwercamacaro! 
I have one additional question I missed from the last review.

Also can you remove the `ProcessorStateManager.assignToStreamThread` method? it's not used anymore.

The `ProcessorStateManagerTest` could use better coverage of `registerStartupStateStores` and `initializeStoreOffsetsFromCheckpoint`

bbejeck commented:
@eduwercamacaro - you can ignore my comments on testing - we can do as a follow-up PR

eduwercamacaro committed:
Merge branch 'trunk' into refactor-startup-store-initialization

eduwercamacaro committed:
Set a log prefix for startup stores during initialization

bbejeck commented:
@eduwercamacaro the failures are related - I think [addressing this comment](https://github.com/apache/kafka/pull/20749#discussion_r2453535667) should fix the integration tests.

eduwercamacaro committed:
Properly assign a StreamThread to a ProcessorStateManager created during startup store pre-initialization

eduwercamacaro committed:
Implement preInit method for InMemory stores

eduwercamacaro committed:
Revert "Implement preInit method for InMemory stores"

This reverts commit b4cb0ff7bba4218251cdf4490edca0b2afd665d0.

eduwercamacaro committed:
wip

eduwercamacaro committed:
Refactor: Open and close state stores during pre-existing stores initialization

eduwercamacaro committed:
Refacttor: remove startup stores from state manager

eduwercamacaro committed:
Revert changes in StandbyTaskCreator

eduwercamacaro committed:
rename methods

eduwercamacaro committed:
Fix unit tests

eduwercamacaro committed:
Remove unused method

eduwercamacaro commented:
@bbejeck I modified this PR according to our last discussion about the solution for [KAFKA-19434](https://issues.apache.org/jira/browse/KAFKA-19434)

We basically open the stores found in the state directory during initialization and then close them. 

cc: @mjsax

eduwercamacaro committed:
Merge branch 'trunk' into refactor-startup-store-initialization

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -237,35 +252,18 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
+                    } finally {
+                        // Make sure the state manager writes the local checkpoint file before closing the stores
+                        // This will be replaced in the future when removing the checkpoint file dependency.
+                        temporaryStateManager.checkpoint();
```
Same as above `temporaryStateManager.checkpoint()` can throw a `ProcessorStateException` but we don't handle it here, meaning `close()` is not called.  So do we consider this a fatal exception and allow it to bubble up and kill the main thread? If not, I'm guessing we'll need to catch this exception and do something.  Either way let's add a couple of comments on the handling

eduwercamacaro replied:
I moved this logic inside the try/catch block to ensure that we always call the `close()` method. Any exception from the `close()` method is considered fatal and will break the main thread.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -237,35 +252,18 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
-                        task.closeDirty();
+                        StateManagerUtil.registerStateStores(log, threadLogPrefix, subTopology, temporaryStateManager, this, initContext);
+                    } catch (final TaskCorruptedException tce) {
+                        log.warn("Failed to register startup state stores for task {}: {}", id, tce.getMessage());
```
`registerStateStores` also can throw a `StreamsException`, `LockException` , or  a `ProcessorStateException` should we handle these exceptions as well? If not add some comments on the reasoning

eduwercamacaro replied:
Good point.
I added some comments and JavaDoc for this method.
Those exceptions are considered fatal and will break the StreamThread. We only handle `TaskCorruptedException` by logging and continuing with the initialization as this exception will be handled by the StreamThread later in the first assignment.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -237,35 +252,18 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
+                        // Make sure the state manager writes the local checkpoint file before closing the stores
+                        // This will be replaced in the future when removing the checkpoint file dependency.
+                        temporaryStateManager.checkpoint();
+                        temporaryStateManager.close();
```
I think we also need to add `StateDirectory.unlock` here  too.

eduwercamacaro replied:
Currently, it is not needed because we keep the lock on the main thread until it is assigned to a SthreadThread. TaskManager changes the ownership of the Task using 'StateDirectory.removeStartupState'.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -237,35 +252,18 @@ public void initializeStartupTasks(final TopologyMetadata topologyMetadata,
+                        temporaryStateManager.checkpoint();
+                        temporaryStateManager.close();
                     }
+                    tasksForLocalState.put(id, new StartupState(id, subTopology, temporaryStateManager));
```
`temporaryStateManager` is closed above  line 264 but it's passed here to `StartupState`.
It's never accessed from `StartupState` .  With the new work is `StartupState` required or can we get away with `Set<TaskId>` ? Or is it going to be needed for subsequent work?

eduwercamacaro replied:
Yep. Excellent observation. 
You are right, we can use `Set<TaskId>`` instead. 
I completely removed the `StartupState` class and the process to close the stores because we are immediately closing them.

bbejeck reviewed (commented):
Thanks for the PR, @eduwercamacaro! Overall, it looks good with few comments.

Also, ` ProcessorStateManager.assignToStreamThread` on line 237  is unused; can we get rid of it?

eduwercamacaro committed:
streams: simplify startup state handling and locking

- track startup task ids via a simple Set and unlock them on close
- remove startup-task handoff/close APIs and related tests
- drop unused standby-thread assignment helper in

eduwercamacaro reviewed (commented):
@bbejeck  Thanks for your comments 
I pushed a new commit with the suggestions.

eduwercamacaro committed:
Merge branch 'trunk' into refactor-startup-store-initialization

eduwercamacaro committed:
- Rename unit tests related to startup states
- Remove startup state on statedir auto cleanup

bbejeck reviewed (approved):
Thanks @eduwercamacaro! LGTM just waiting for the build to complete

bbejeck commented:
Merged #20749 into trunk

omkreddy commented:
looks like build is failing after this commit

bbejeck commented:
> looks like build is failing after this commit

Thanks @omkreddy! I've pushed a fix we should be good now
